subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_uart.c

```diff
@@ -0,0 +1,307 @@
+
+		if (new_len > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER) {
+			/* Bad data error, set the state and wait for timeout */
+			hc_uart->state = UART_HOST_CMD_RX_BAD;
+			return;
+		}
+
+		hc_uart->rx_ctx->len = new_len;
+
+		if (hc_uart->rx_ctx->len >= sizeof(struct ec_host_cmd_request_header)) {
+			/* Buffer has request header. Check header and get data_len */
+			hc_uart->expected_len = request_expected_size(
+				(struct ec_host_cmd_request_header *)hc_uart->rx_ctx->buf);
+
+			if (hc_uart->expected_len == 0 ||
+			    hc_uart->expected_len > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER) {
+				/* Invalid expected size, set the state and wait for timeout */
+				hc_uart->state = UART_HOST_CMD_RX_BAD;
+			} else if (hc_uart->rx_ctx->len == hc_uart->expected_len) {
+				/* Don't wait for overrun, because it is already done
+				 * in a UART driver.
+				 */
+				(void)k_work_cancel_delayable(&hc_uart->timeout_work);
+
+				/* Disable receiving to prevent overwriting the rx buffer while
+				 * processing. Enabling receiving to a temporary buffer to detect
+				 * unexpected transfer while processing increases average handling
+				 * time ~40% so don't do that.
+				 */
+				uart_rx_disable(hc_uart->uart_dev);
+
+				/* If no data more in request, packet is complete. Start processing
+				 */
+				hc_uart->state = UART_HOST_CMD_PROCESSING;
+
+				k_sem_give(&hc_uart->rx_ctx->handler_owns);
+			} else if (hc_uart->rx_ctx->len > hc_uart->expected_len) {
+				/* Overrun error, set the state and wait for timeout */
+				hc_uart->state = UART_HOST_CMD_RX_OVERRUN;
+			}
+		}
+		break;
+	case UART_RX_BUF_REQUEST:
+		/* Do not provide the second buffer, because we reload DMA after every packet. */
+		break;
+	case UART_TX_DONE:
+		/* Receiving is already enabled in the send function. */
+		hc_uart->state = UART_HOST_CMD_READY_TO_RX;
+		break;
+	case UART_RX_STOPPED:
+		LOG_ERR("UART HOST CMD ERROR: Receiving data stopped");
+		break;
+	default:
+		break;
+	}
+}
+
+static int ec_host_cmd_uart_init(const struct ec_host_cmd_backend *backend,
+				 struct ec_host_cmd_rx_ctx *rx_ctx, struct ec_host_cmd_tx_buf *tx)
+{
+	int ret;
+	struct ec_host_cmd_uart_ctx *hc_uart = backend->ctx;
+
+	hc_uart->state = UART_HOST_CMD_STATE_DISABLED;
+
+	if (!device_is_ready(hc_uart->uart_dev)) {
+		return -ENODEV;
+	}
+
+	/* UART backend needs rx and tx buffers provided by the handler */
+	if (!rx_ctx->buf || !tx->buf) {
+		return -EIO;
+	}
+
+	hc_uart->rx_ctx = rx_ctx;
+	hc_uart->tx_buf = tx;
+
+	k_work_init_delayable(&hc_uart->timeout_work, rx_timeout);
+	uart_callback_set(hc_uart->uart_dev, uart_callback, hc_uart);
+	ret = uart_rx_enable(hc_uart->uart_dev, hc_uart->rx_ctx->buf,
+			     CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER, UART_OVERRUN_TIMEOUT_US);
+
+	hc_uart->state = UART_HOST_CMD_READY_TO_RX;
+
+	return ret;
+}
+
+static int ec_host_cmd_uart_send(const struct ec_host_cmd_backend *backend)
+{
+	struct ec_host_cmd_uart_ctx *hc_uart = backend->ctx;
+	int ret;
+
```

Comment:
nit: We expect the state to be `UART_HOST_CMD_PROCESSING` here. I would check it here, but you can decide.